source/adios2/core/VariableBase.cpp

```diff
@@ -72,16 +73,21 @@ MemorySpace VariableBase::DetectMemorySpace(const void *ptr)
 
 MemorySpace VariableBase::GetMemorySpace(const void *ptr)
 {
+#ifdef ADIOS2_HAVE_GPU_SUPPORT
     if (m_MemSpaceDetected == MemorySpace::Detect)
         m_MemSpaceDetected = DetectMemorySpace(ptr);
     return m_MemSpaceDetected;
+#endif
+    return MemorySpace::Host;
 }
 
 void VariableBase::SetMemorySpace(const MemorySpace mem)
 {
     m_MemSpaceRequested = mem;
     // reset the detected memory space
+#ifdef ADIOS2_HAVE_GPU_SUPPORT
     m_MemSpaceDetected = MemorySpace::Detect;
+#endif
```

Comment:
```suggestion
```

In my opinion we do not need to detect the `MemorySpace` for cases when the user set it to anything other than `MemorySpace::Detect`. This is for simplicity on the API. It might seem strange to the user to `SetMemorySpace` to one value but then `GetMemorySpace` to another, I think it is better to let it error if the user use the wrong parameters.

Reply:
I agree that it's not the cleanest way to do this. `SetMemorySpace` gives a hint to what kind of buffers to expect. 
`GetMemorySpace` returns what we are actually receiving. The `MemSpaceRequested` variable is the one requested by the user, but not necessarily the one used internally. `MemSpaceDetected` is what we use internally to decided what backend to use. This differs from what is set by the user only if the application is using `Kokkos::View`s. If a user sets the memory space to `MemorySpace::Host` but then puts a CUDA Kokkos view the detected memory space will be `MemorySpace::CUDA` but at the next loop we want to reset it to what the user requested.

I agree it is a bit of a pain, so I would be ok with giving this up and throwing an exception if the `Kokkos::View` memory space does not match what the users sets.